src/multiplicity_vectors.lean

```diff
@@ -0,0 +1,645 @@
+begin
+  rw [pnat.coprime, disjoint_iff, ← factor_finsupp_gcd_eq_inf_factor_finsupps,
+    finsupp.bot_eq_zero],
+  rw ← factor_finsupp_one, split; intro h, rw h,
+  apply pnat.prime_finsupp_equiv.injective, simpa,
+end
+
+lemma coprime_iff_disjoint_supports {m n : ℕ+} :
+  m.coprime n ↔ disjoint (factor_finsupp m).support (factor_finsupp n).support :=
+begin
+  rw coprime_iff_disjoint_factor_finsupps, rw finsupp.disjoint_iff,
+  unfold disjoint, repeat {rw le_bot_iff},
+  split; intro h; rw ← h; ext; simp, --WHY DOESN'T REFL WORK???
+end
+
+end basic_number_theory_definitions
+
+@[instance]
+def prime_finsupp_coe_pnat : has_coe (nat.primes →₀ ℕ) (ℕ+ →₀ ℕ) :=
+{ coe := finsupp.map_domain coe }
+
+-- Just wraps to_multiset in the prime_multiset type for the next lemma to typecheck
+def finsupp.to_prime_multiset (f : nat.primes →₀ ℕ) : prime_multiset := f.to_multiset
+
+lemma coe_pnat_commute_to_multiset {f : nat.primes →₀ ℕ} :
+(↑f : ℕ+ →₀ ℕ).to_multiset =  prime_multiset.to_pnat_multiset f.to_prime_multiset :=
+begin
+  unfold prime_multiset.to_pnat_multiset, unfold finsupp.to_prime_multiset,
+  rw finsupp.to_multiset_map, refl
+end
+
+lemma prod_pow_factor_finsupp (n : ℕ+) :
+  (factor_finsupp n).prod_pow = n :=
+begin
+  rw prime_finsupp.prod_pow_eq_prod_to_multiset,
+  rw factor_finsupp_to_multiset_eq_factor_multiset,
+  rw pnat.prod_factor_multiset
+end
+
+lemma factor_finsupp_prod_pow (f : prime_finsupp) :
+factor_finsupp (f.prod_pow) = f :=
+begin
+  unfold factor_finsupp, conv_rhs {rw ← f.to_multiset_to_finsupp},
+  rw ← prime_multiset.factor_multiset_prod f.to_multiset,
+  rw prime_finsupp.prod_pow_eq_prod_to_multiset
+end
+
+lemma factor_finsupp_inj : function.injective factor_finsupp :=
+begin
+  unfold function.injective, intros a b h,
+  rw [← prod_pow_factor_finsupp a, ← prod_pow_factor_finsupp b, h]
+end
+
+section prime_powers
+
+variables {p : nat.primes} {n : ℕ+} {k : ℕ}
+
+@[simp]
+lemma factor_finsupp_prime : factor_finsupp ↑p = finsupp.single p 1 :=
+begin
+  apply finsupp.equiv_multiset.injective,
+  change finsupp.to_multiset (factor_finsupp ↑p) = finsupp.to_multiset (finsupp.single p 1),
+  rw [finsupp.to_multiset_single, factor_finsupp_to_multiset_eq_factor_multiset,
+    pnat.factor_multiset_of_prime, prime_multiset.of_prime],
+  simp
+end
+
+@[simp]
+lemma factor_finsupp_pow : factor_finsupp (n ^ k) = k • factor_finsupp n :=
+begin
+  induction k, simp,
+  rw [pow_succ, nat.succ_eq_add_one, add_smul, ← k_ih, mul_comm], simp
+end
+
+@[simp]
+lemma prime_finsupp.smul_eq_smul {b : ℕ} :
+  (k • finsupp.single p b : prime_finsupp) = (k • finsupp.single p b : nat.primes →₀ ℕ) :=
+begin
+  rw nat.smul_def,
+  induction k, simp, rw [succ_nsmul, nat.succ_eq_add_one, k_ih, add_smul, add_comm], simp
+end
+
+@[simp]
+lemma nat.smul_one : k • 1 = k := by { rw nat.smul_def, simp }
+
+@[simp]
+lemma factor_finsupp_pow_prime : factor_finsupp (p ^ k) = finsupp.single p k := by simp
+
+end prime_powers
+
+section pnat_facts
```

Comment:
This section looks really good to me. It can be PRd to `data/pnat/*`.
I think all of it can go in `xgcd.lean` in that directory, but if you find better places, go ahead.

Reply:
https://github.com/leanprover-community/mathlib/pull/3274